src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs

```diff
@@ -741,35 +1068,91 @@ public enum RoundType
             Down
         }
 
-        // Char is used for PA string escaping 
-        public static FormulaValue RoundUp(IRContext irContext, NumberValue[] args)
+        public static FormulaValue RoundUp(IRContext irContext, FormulaValue[] args)
         {
-            var numberArg = args[0].Value;
-            var digitsArg = args[1].Value;
+            double digits;
 
-            var x = Round(numberArg, digitsArg, RoundType.Up);
-            return new NumberValue(irContext, x);
+            if (args.Length == 2 && args[1] is NumberValue numberDigs)
+            {
+                digits = numberDigs.Value;
+            }
+            else
+            {
+                return CommonErrors.UnreachableCodeError(irContext);
+            }
+
+            if (args[0] is NumberValue num)
+            {
+                return RoundFloat(irContext, num, digits, RoundType.Up);
+            }
+            else if (args[0] is DecimalValue dec)
+            {
+                return RoundDecimal(irContext, dec, digits, RoundType.Up);
+            }
+            else
+            {
+                return CommonErrors.UnreachableCodeError(irContext);
+            }
         }
 
-        public static FormulaValue RoundDown(IRContext irContext, NumberValue[] args)
+        public static FormulaValue RoundDown(IRContext irContext, FormulaValue[] args)
         {
-            var numberArg = args[0].Value;
+            double digits;
 
-            // RoundDown is also used for Trunc which has an optional digits argument
-            var digitsArg = args.Length == 1 ? 0 : args[1].Value;
+            // Trunc uses RoundDown as the implementation, and Trunc's second argument is optional
+            if (args.Length == 2 && args[1] is NumberValue numberDigs)
+            {
+                digits = numberDigs.Value;
+            }
+            else if (args.Length == 1)
+            {
+                digits = 0;
+            }
+            else
+            {
+                return CommonErrors.UnreachableCodeError(irContext);

```

Comment:
>eturn CommonErrors.UnreachableCodeError(irContext); [](http://example.com/codeflow?start=17&length=51)

would we reach that point if args.Length == 2 && args[1] is DecimalValue ?

Reply:
Correct.  CheckTypes ensures that the second argument is Number, as does Library.cs.